R/scale-.r

```diff
@@ -810,6 +816,399 @@ ScaleDiscrete <- ggproto("ScaleDiscrete", Scale,
   }
 )
 
+#' @rdname ggplot2-ggproto
+#' @format NULL
+#' @usage NULL
+#' @export
+ScaleBinned <- ggproto("ScaleBinned", Scale,
+  range = continuous_range(),
+  na.value = NA_real_,
+  oob = squish,
+  n.breaks = NULL,
+  right = TRUE,
+  after_stat = FALSE,
+  show.limits = FALSE,
+
+  is_discrete = function() FALSE,
+
+  train = function(self, x) {
+    if (!is.numeric(x)) {
+      stop("Binned scales only support continuous data", call. = FALSE)
+    }
+
+    if (length(x) == 0) return()
+    self$range$train(x)
+  },
+
+  transform = function(self, x) {
+    new_x <- self$trans$transform(x)
+    if (any(is.finite(x) != is.finite(new_x))) {
+      type <- if (self$scale_name == "position_c") {
+        "continuous"
+      } else if (self$scale_name == "position_d") {
+        "discrete"
+      } else {
+        "binned"
+      }
+      axis <- if ("x" %in% self$aesthetics) "x" else "y"
+      warning("Transformation introduced infinite values in ", type, " ", axis, "-axis", call. = FALSE)
+    }
+    new_x
+  },
+
+  map = function(self, x, limits = self$get_limits()) {
+    if (self$after_stat) {
+      x
+    } else {
+      breaks <- self$get_breaks(limits)
+
+      x_binned <- cut(x, c(limits[1], breaks, limits[2]), labels = FALSE,
+                      include.lowest = TRUE, right = self$right)
+
+      if (!is.null(self$palette.cache)) {
+        pal <- self$palette.cache
+      } else {
+        pal <- self$palette(length(self$breaks) + 1)
+        self$palette.cache <- pal
+      }
+
+      pal[x_binned]
+    }
+  },
+
+  dimension = function(self, expand = c(0, 0, 0, 0)) {
+    expand_range4(self$get_limits(), expand)
+  },
+
+  get_breaks = function(self, limits = self$get_limits()) {
+    if (self$is_empty()) return(numeric())
+
+    limits <- self$trans$inverse(limits)
+
+    if (is.null(self$breaks)) {
+      return(NULL)
+    } else if (identical(self$breaks, NA)) {
+      stop("Invalid breaks specification. Use NULL, not NA", call. = FALSE)
+    } else if (is.waive(self$breaks)) {
+      if (!is.null(self$n.breaks)) {
+        old_n <- get("n", environment(self$trans$breaks))
```

Comment:
This is moderately horrendous. Why do we need to write into this environment?

Reply:
That is putting it mildly 🙂 It is atrocious and it won't be part of the merge. I put it in as a hack to allow setting number of breaks in the transformation object, but this has been fixed in scales (also why we need a scale release soon)

My plan is to inspect the formals of the `breaks` method and, if `n` is present, use it. In this way we don't need to rely on updated transformation objects